### 4.7 Deriving new variables<br>
-if, then<br>
-apply label<br>
-.loc

In [13]:
###import libraries
import pandas as pd
import numpy as np
import os as os

In [14]:
path =r'/Users/margoshogren/Desktop/1 Margo-CF Folder/4 Python/4-2023 Instacart Basket Analysis'

### Import orders_prods_merged.pkl

In [15]:
ords_prods_merge = pd.read_pickle(os.path.join(path,'2 Data','prepared data','orders_products_merged.pkl'))

### Question(1) Create 'price_label' and 'busiest_day' columns

#### First try to create a subset and if,then statement

In [16]:
df_subset = ords_prods_merge[:100000]

In [17]:
df = ords_prods_merge

In [18]:
def price_label(row):
    
  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'


### apply the price_label statement

In [19]:
df['price_range'] = df.apply(price_label, axis = 1)

In [54]:
ords_prods_merge = ords_prods_merge.drop(columns = ['price_range'])

#### Observation: Python suggest .loc method to be more efficient

In [20]:
# Create conditions with loc for high, mid, low range; python says the .loc will be more a more efficient method

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [21]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [22]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

### Frequency count of the 3 groups in 'price_range_loc'

In [23]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product     21861558
Low-range product     10130750
High-range product      412551
Name: price_range_loc, dtype: int64

#### Observation: Frequency count shows 'Mid-range' the most frequent

### Create a For Loop for 'Busiest day, Least busy, Regularly busy' ; put result on 'result'

In [24]:
result = []
for value in df['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

#### Create a new column = result

In [26]:
df['busiest_day'] = result

In [27]:
df['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

#### Compare above frequency count to column frequency count below

In [28]:
df['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#### Observation: Shows busiest day '0' (Saturday), '4' (Tuesday) least busy

### Question(2) Change busiest_day column to be 'busiest days' = 2 busiest days, add 2 slowest days

In [29]:
result = []
for value in df['orders_day_of_week']:
    if value <= 1:
        result.append('Busiest days')
    elif (value == 3 or value == 4):
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [30]:
df['Busiest_days'] = result

In [53]:
ords_prods_merge = ords_prods_merge.drop(columns = ['busiest_day'])

### Question(3) Check the values of the new result

In [31]:
df['Busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Least busy         7624336
Name: Busiest_days, dtype: int64

#### Observations <br>
**Busiest days:**<br>
(0) 6204182<br>
(1) 5660230<br>

**Least busy:**<br>
(3) 3840534<br>
(4) 3783802<br>
*result as intended*

### Question(4)
*Identify the busiest hours of the day.*<br> 
Rather than by hour, they want periods of time labeled **“Most orders,” “Average orders,” and “Fewest orders.”**<br>  Create a new column containing these labels called **“busiest_period_of_day.”**

In [55]:
ords_prods_merge.columns

Index(['Unnamed: 0_x', 'product_id', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Unnamed: 0_y', 'order_id', 'user_id',
       'order_number', 'orders_day_of_week', 'order_hour_of_day',
       'days_since_prior_order', 'add_to_cart_order', 'reordered', '_merge',
       'price_range_loc', 'Busiest_days', 'busiest_period_of_day'],
      dtype='object')

#### Identify busiest hours of the day

In [33]:
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

#### Observation:
busiest hours (> 250k count) is 10am to 4pm;<br>
Least busy (< 40k count) is 12am to 6am.

In [34]:
result = []
for value in df['order_hour_of_day']:
    if value > 9:
        result.append('Most orders')
    elif (value < 6):
        result.append('Fewest orders')
    else:
        result.append('Average orders')

In [35]:
df['busiest_period_of_day'] = result

In [41]:
ords_prods_merge.describe()

Unnamed: 0_x    product_id      aisle_id  department_id        prices  \
count  3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07   
mean   2.560037e+04  2.559866e+04  7.119612e+01   9.919792e+00  7.790180e+00   
std    1.408555e+04  1.408400e+04  3.821139e+01   6.281485e+00  4.242125e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00   
25%    1.354400e+04  1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00   
50%    2.530300e+04  2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00   
75%    3.795100e+04  3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01   
max    4.969200e+04  4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01   

       Unnamed: 0_y      order_id       user_id  order_number  \
count  3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean   1.709858e+06  1.710745e+06  1.029372e+05  1.714230e+01   
std    9.864575e+05  9.872988e+05  5.946610e+04  1.753532e+01   
min    0.000000e+00  2.000000e+00  1.000000e+00  1.000000e+00   
25%    8.556370e+05  8.559470e+05  5.142200e+04  5.000000e+00   
50%    1.709176e+06  1.711049e+06  1.026160e+05  1.100000e+01   
75%    2.565919e+06  2.565499e+06  1.543890e+05  2.400000e+01   
max    3.421081e+06  3.421083e+06  2.062090e+05  9.900000e+01   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  \
count        3.240486e+07       3.240486e+07            3.032876e+07   
mean         2.738867e+00       1.342515e+01            1.110408e+01   
std          2.090077e+00       4.246380e+00            8.779064e+00   
min          0.000000e+00       0.000000e+00            0.000000e+00   
25%          1.000000e+00       1.000000e+01            5.000000e+00   
50%          3.000000e+00       1.300000e+01            8.000000e+00   
75%          5.000000e+00       1.600000e+01            1.500000e+01   
max          6.000000e+00       2.300000e+01            3.000000e+01   

       add_to_cart_order     reordered  
count       3.240486e+07  3.240486e+07  
mean        8.352547e+00  5.895873e-01  
std         7.127071e+00  4.919087e-01  
min         1.000000e+00  0.000000e+00  
25%         3.000000e+00  0.000000e+00  
50%         6.000000e+00  1.000000e+00  
75%         1.100000e+01  1.000000e+00  
max         1.450000e+02  1.000000e+00

In [52]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 50]

Empty DataFrame
Columns: [Unnamed: 0_x, product_id, product_name, aisle_id, department_id, prices, Unnamed: 0_y, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, add_to_cart_order, reordered, _merge, price_range, price_range_loc, busiest_day, Busiest_days, busiest_period_of_day]
Index: []

[0 rows x 21 columns]

### Question(5) Print the frequency for this new column.

In [36]:
df['busiest_period_of_day'].value_counts(dropna = False)

Most orders       26454663
Average orders     5353868
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

#### Observation:
Most orders (>250k per hour) are from **10am-4pm**;<br>
Fewest orders (<40k per hour) are from 12am-6am;<br> 
Average orders are after 4pm till 11:59pm.

In [57]:
ords_prods_merge.columns

Index(['Unnamed: 0_x', 'product_id', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Unnamed: 0_y', 'order_id', 'user_id',
       'order_number', 'orders_day_of_week', 'order_hour_of_day',
       'days_since_prior_order', 'add_to_cart_order', 'reordered', '_merge',
       'price_range_loc', 'Busiest_days', 'busiest_period_of_day'],
      dtype='object')

In [56]:
#### Export file as pickle
ords_prods_merge.to_pickle(os.path.join(path,'2 Data','prepared data','orders_products_w-newcols.pkl'))

#### *End of 4.7 Deriving new variables notebook.